virtualizarr/readers/common.py

```diff
@@ -1,121 +1,107 @@
+def construct_fully_virtual_dataset(
+    virtual_vars: Mapping[str, xr.Variable],
+    coord_names: Iterable[str] | None = None,
+    attrs: dict[str, Any] | None = None,
+) -> xr.Dataset:
+    """Construct a fully virtual Dataset from constituent parts."""
+
+    data_vars, coords = separate_coords(
+        vars=virtual_vars,
+        indexes={},  # we specifically avoid creating any indexes yet to avoid loading any data
+        coord_names=coord_names,
+    )
 
-    Relies on xr.open_dataset and its auto-detection of filetypes to find the correct installed backend.
-    """
+    vds = xr.Dataset(
+        data_vars=data_vars,
+        coords=coords,
+        attrs=attrs,
+    )
 
-    if loadable_variables == [] and indexes == {}:
-        # no need to even attempt to open the file using xarray
-        return {}, {}
+    # TODO we should probably also use vds.set_close() to tell xarray how to close the file we opened
 
-    # TODO We are reading a bunch of stuff we know we won't need here, e.g. all of the data variables...
-    # TODO It would also be nice if we could somehow consolidate this with the reading of the kerchunk references
-    # TODO Really we probably want a dedicated backend that iterates over all variables only once
-    # TODO See issue #124 for a suggestion of how to avoid calling xarray here.
+    return vds
+
+
+# TODO reimplement this using ManifestStore (GH #473)
+def replace_virtual_with_loadable_vars(
+    fully_virtual_dataset: xr.Dataset,
+    filepath: str,  # TODO won't need this after #473 because the filepaths will be in the ManifestStore
+    group: str | None = None,
+    loadable_variables: Iterable[Hashable] | None = None,
+    decode_times: bool | None = None,
+    indexes: Mapping[str, xr.Index] | None = None,
+    reader_options: Optional[dict] = None,
+) -> xr.Dataset:
+    if indexes is not None:
+        raise NotImplementedError()
 
     fpath = _FsspecFSFromFilepath(
         filepath=filepath, reader_options=reader_options
     ).open_file()
 
-    # fpath can be `Any` thanks to fsspec.filesystem(...).open() returning Any.
-    ds = open_dataset(
+    # TODO replace with only opening specific variables via `open_zarr(ManifestStore)` in #473
+    loadable_ds = xr.open_dataset(
         fpath,  # type: ignore[arg-type]
-        drop_variables=drop_variables,
         group=group,
         decode_times=decode_times,
     )
 
-    if indexes is None:
-        # add default indexes by reading data from file
-        # but avoid creating an in-memory index for virtual variables by default
-        indexes = {
-            name: index
-            for name, index in ds.xindexes.items()
-            if name in loadable_variables
-        }
-    elif indexes != {}:
-        # TODO allow manual specification of index objects
-        raise NotImplementedError()
+    var_names_to_load: list[Hashable]
+    if isinstance(loadable_variables, list):
+        var_names_to_load = list(loadable_variables)
```

Comment:
why do you need to cast this to a list if the code is only run on list instances?

Reply:
Because I made the input type an `Iterable`, not a `list`